### PR TITLE
Beta Version of Blender Exporter

### PR DESCRIPTION
**Pull request by ProjectBarks**
The design of the exporter was to incorporate blender design paradigms while working within available tooling.

# Design
Files follow standard python guidelines in code seperation while placing a larger emphasis on smaller readable files rather than large unreadable files. Files placed in `operators` are executed based tasks that influence the user interface...
- `oauth` starts a server to catch the token and shuts down once received
- `token` once the token is received caching occurs while forcing a update
- `upoad_gltf` exports a base gltf and uploads it after receiving required tokens

The `panels` are minimal implementation of UI used primarily as a user facing endpoint. `__init__` handles registration and loading preferences. `cache` is a simple way of dumping preferences into json. `server` is a dumb http server that returns `index.html` and shuts down on the first token it receives. All tokens and basic settings are stored in `globals.py`

**IGNORE VENDORS** - It is required dependencies for blender but not automatically installed

# Todo
- Add Menu File export
- Add Export Options

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,15 @@
 # ion-blender-exporter
 Blender add-on for uploading and tiling models with Cesium ion. https://cesium.com
+
+## Installation
+1. Download the project as a zip [here](https://github.com/AnalyticalGraphicsInc/ion-blender-exporter/archive/beta.zip)
+2. Go to `File > Preferences > Add-ons`
```
Comment (mramato) [inline README.md]: I had to go to `Edit -> Preferences -> Add-ons`.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,15 @@
 # ion-blender-exporter
 Blender add-on for uploading and tiling models with Cesium ion. https://cesium.com
+
+## Installation
+1. Download the project as a zip [here](https://github.com/AnalyticalGraphicsInc/ion-blender-exporter/archive/beta.zip)
```
Comment (mramato) [inline README.md]: This is OK for now, but we'll need to figure out the release process, I assume we will eventually link to the latest official releases taged with GitHub releases.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,15 @@
...
+2. Go to `File > Preferences > Add-ons`
+3. Enable Add-on
+
+## Todo
```
Comment (mramato) [inline README.md]: We don't keep TODO lists in the README.  You can move these into the PR description for now (we normally use task lists for this kind of stuff).  For stuff not part of the PR, we should write up new issues.

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,130 @@
+###OSX###
```
Comment (mramato) [inline .gitignore]: We use minimal ignore lists to make them easier to maintain.  Please remove everything here that isn't necessary for this project.

File: __init__.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+bl_info = {
+    "name": "Cesium Ion",
+    "description": "Cesium Ion Exporting Tool",
+    "author": "",
```
Comment (mramato) [inline __init__.py]: Author should be `Analytical Graphics, Inc.`

**Comment by mramato** [comment]
Haven't looked at the code yet, but here's some general top-level UX feedback from my first few minutes (some of these we may have already discussed)

1. Why have a separate `Authorize` step, can't we just authorize the first time they hit the export button?
2. Rather than have a modeless dialog that they have to fill in, why not a modal that pops up with whatever options are available (this will also make it easy to pre-populate it with defaults from the Blender file we are exporting.
3. Similar to 2, from this tutorial https://docs.blender.org/manual/en/dev/advanced/scripting/addon_tutorial.html, it looks like we could just either insert ourselves in the `File` menu, or add a new top-level Cesium ion menu (though I think it makes more sense to `File->Export->Cesium ion` or `File->Upload-> Cesium ion` if we can't add to the sub-menu.
4. We need to have a radio button that let's uses choose between `3D Model` and `3D Capture` so that we can set the equivalent option in the REST API (there shouldn't be a default, make them choose).

File: README.md (hunk 2 above)
Comment (ProjectBarks) [reply README.md]: Yep

**Comment by mramato** [comment]
Is `vendor` required to be the directory name or otherwise standardized in python?  In all Cesium codebases, we always call this `ThirdParty`

File: README.md (hunk 3 above)
Comment (ProjectBarks) [reply README.md]: Yeah I thought it would help contextualize where I am taking the codebase

File: __init__.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+bl_info = {
+    "name": "Cesium Ion",
```
Comment (mramato) [inline __init__.py]: `ion` is always lower-cased.  The only exception is code symbols for readability `IonClass`, `CesiumIonClass`, etc..

File: __init__.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    "description": "Cesium Ion Exporting Tool",
```
Comment (mramato) [inline __init__.py]: Perhaps something like `Upload and tile models with Cesium ion. https://cesium.com` We always have to be mindful when communicating information to the user.  I assume this is just plain text?  (No markdown or anything?)

**Comment by mramato** [comment]
This is coming along nicely.  Great work @ProjectBarks! Since I asked for some non-trivial changes I'm going to stop reviewing for now.  Let me know if you have any additional questions and just bump when you think it's ready for another look.

File: README.md (hunk 3 above)
Comment (mramato) [reply README.md]: That's a valid approach, just not how we use READMEs throughout Cesium. At long as it's cleaned up before we merge, that's fine.

File: README.md (hunk 2 above)
Comment (ProjectBarks) [reply README.md]: TODO: Switch this to using releases instead & Write release documentation

**Comment by ProjectBarks** [comment]
Ready for review

**Comment by ProjectBarks** [comment]
@mramato ready for review

**Comment by mramato** [comment]
Taking a look now.

File: panels/export.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        csm_export = context.scene.csm_export
+        csm_user_len = len(context.window_manager.csm_user.token)
+
+        if csm_user_len > 0:
+            layout = self.layout
+            scene = context.scene
+
+            layout.prop(scene.csm_export, "name")
+            if len(csm_export.name) <= 0:
+                layout.label(text="Set name before export...")
```
Comment (mramato) [inline panels/export.py]: Let's just change this label to `Name is required`.

File: panels/export.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+            layout.prop(scene.csm_export, "description")
+            layout.prop(scene.csm_export, "attribution")
+            layout.separator()
+            layout.prop(scene.csm_export, "source_type")
+            if csm_export.source_type is "":
+                layout.label(text="Set source type before export...")
```
Comment (mramato) [inline panels/export.py]: For UX purposes, let's refer to it as `Model type` (since Blender only works in models while source is more general to ion's multi-domain usage).  Also change this to `Model type is required`.

**Comment by mramato** [comment]
This turned out super nice, awesome work @ProjectBarks.  I think I just have some minor stuff that I'm going through now.

**Comment by ProjectBarks** [comment]
@mramato made those naming changes

File: panels/export.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
+from bpy.props import *
+from bpy.types import (Panel, PropertyGroup)
+
+from ..operators import (ExportUploadOperator, OAuthOperator)
+from ..globals import APP_CATEGORY
+
+SOURCE_TYPES = [
+    ("3D_MODEL", "3D Model", "Models designed through blender", "FILE_3D", 1),
+    ("3D_CAPTURE", "3D Capture",
+     "Model created through photogrammetry processes", "CAMERA_DATA", 2)
```
Comment (mramato) [inline panels/export.py]: > Model created through photogrammetry processes

Just realized the doc in the REST API is pretty bad.  Let's change this to something like:

Model created through 3D capture such as scanning, LIDAR, photogrammetry, etc...

**Comment by mramato** [comment]
I'm guessing the answer is no, but is there way to expand the default width of the tab?

**Comment by mramato** [comment]
Can we show some sort of feedback if they click `Upload to ion` without entering a name?  Mobile apps shake, or maybe hide the required messages until there is an actual error?  What does Blender's UX allow for?

File: panels/export.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        name="Description",
+        description="Give an overview of the contents of the data")
+    attribution: StringProperty(
+        name="Attribution",
+        description="Acknowledge the particular author, artist, or coporation")
+    source_type: EnumProperty(name="Model Type",
+                              description="Type of data being uploaded",
+                              items=lambda self, context: SOURCE_TYPES)
+    webp_textures: BoolProperty(
+        name="Convert to WebP",
```
Comment (mramato) [inline panels/export.py]: I think this is confusing since it doesn't say what will be WebP.  `Use WebP images` is what the ion UX has, so let's go with that.

File: operators/upload_gltf.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+            self._operator.report({"INFO"}, f"Progress Upload {percentage}")
+
+            if self._seen_so_far >= self._size:
+                wm.progress_end()
+
+
+class ExportUploadOperator(Operator):
+    bl_label = "gltf to Cesium uploader"
+    bl_idname = f"{APP_OPERATOR_PREFIX}.upload_gltf"
+    bl_description = "Uploads \".glb\" exports to Cesium ion"
```
Comment (mramato) [inline operators/upload_gltf.py]: This is oddly phrased.  Let's go with `Uploads your model to Cesium ion for 3D Tiling`

**Comment by ProjectBarks** [comment]
> I'm guessing the answer is no, but is there way to expand the default width of the tab?

You guess correctly

> Can we show some sort of feedback if they click `Upload to ion` without entering a name? Mobile apps shake, or maybe hide the required messages until there is an actual error? What does Blender's UX allow for?

I changed the UI so it is more obvious of what is required. The best we can do is show a required error icon for things not complete. Checkout the newest revision

**Comment by mramato** [comment]
Awesome work @ProjectBarks.
